### PR TITLE
Add `ListTileControlAffinity` to `ListTileTheme`

### LINKED ISSUE #146414: Add ListTileControlAffinity into ListTileThemeData
### Use case

I have an case, my design uses this components SwitchListTile, RadioListTile, CheckboxListTile with controlAffinity: ListTileControlAffinity.leading, could you please add property to ListTileThemeData?

### Proposal

```
static final ListTileThemeData listTileThemeData = ListTileThemeData(
    titleTextStyle: Styles.bodyPrimaryMedium,
    contentPadding: EdgeInsets.zero,
    horizontalTitleGap: 4,
    minVerticalPadding: 0,
    titleAlignment: ListTileTitleAlignment.top,
    //TODO: Add more properties
    //controlAffinity: ListTileControlAffinity.leading,
  );
```

### PR DESCRIPTION
**Pull request by hello-coder-xu**
fixes https://github.com/flutter/flutter/issues/146414

--- 

I saw @abikko submitted a PR https://github.com/flutter/flutter/pull/146630, but it was not completed due to CLA and lack of test cases.
I am willing to complete this PR in combination with @abikko's code (I don't know if this is allowed) 

 This PR adds contorlAffinity to ListTileTheme so that [CheckboxListTile], [RadioListTile], [SwitchListTile], and [ExpansionTile] can read and use it 

 For example: If ListTileTheme in Theme sets contorlAffinity, then [CheckboxListTile] can directly use contorlAffinity in ListTileTheme. Of course, if contorlAffinity is also set in [CheckboxListTile], the property in [CheckboxListTile] will be used first.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by hello-coder-xu** [comment]
@TahaTesser  Can you review my PR?

File: packages/flutter/lib/src/material/checkbox_list_tile.dart [hunk 1]
```diff
@@ -518,11 +518,12 @@ class CheckboxListTile extends StatelessWidget {
         );
     }
 
+    final ThemeData theme = Theme.of(context);
+    final ListTileControlAffinity controlAffinity = this.controlAffinity ?? theme.listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
     final (Widget? leading, Widget? trailing) = switch (controlAffinity) {
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/checkbox_list_tile.dart:523]:
```suggestion
    final (Widget? leading, Widget? trailing) = switch (effectiveControlAffinity) {
```

File: packages/flutter/lib/src/material/radio_list_tile.dart [hunk 2]
```diff
@@ -487,13 +487,13 @@ class RadioListTile<T> extends StatelessWidget {
           ),
         );
     }
-
+    final ThemeData theme = Theme.of(context);
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/radio_list_tile.dart:490]:
```suggestion

    final ThemeData theme = Theme.of(context);
```

File: packages/flutter/lib/src/material/radio_list_tile.dart [hunk 3]
```diff
@@ -487,13 +487,13 @@ class RadioListTile<T> extends StatelessWidget {
...
+    final ListTileControlAffinity controlAffinity = this.controlAffinity ?? theme.listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
     Widget? leading, trailing;
     (leading, trailing) = switch (controlAffinity) {
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/radio_list_tile.dart:493]:
```suggestion
    (leading, trailing) = switch (effectiveControlAffinity) {
```

File: packages/flutter/lib/src/material/switch_list_tile.dart [hunk 4]
```diff
@@ -565,14 +565,14 @@ class SwitchListTile extends StatelessWidget {
           ),
         );
     }
-
+    final ThemeData theme = Theme.of(context);
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/switch_list_tile.dart:568]:
```suggestion

    final ThemeData theme = Theme.of(context);
```

File: packages/flutter/lib/src/material/switch_list_tile.dart [hunk 5]
```diff
@@ -565,14 +565,14 @@ class SwitchListTile extends StatelessWidget {
...
+    final ListTileControlAffinity controlAffinity = this.controlAffinity ?? theme.listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
     Widget? leading, trailing;
     (leading, trailing) = switch (controlAffinity) {
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/switch_list_tile.dart:571]:
```suggestion
    (leading, trailing) = switch (effectiveControlAffinity) {
```

File: packages/flutter/lib/src/material/checkbox_list_tile.dart [hunk 6]
```diff
@@ -518,11 +518,12 @@ class CheckboxListTile extends StatelessWidget {
         );
     }
 
+    final ThemeData theme = Theme.of(context);
+    final ListTileControlAffinity controlAffinity = this.controlAffinity ?? theme.listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/checkbox_list_tile.dart:522]:
```suggestion
    final ListTileThemeData listTileTheme = ListTileTheme.of(context);
    final ListTileControlAffinity effectiveControlAffinity = controlAffinity
      ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```

File: packages/flutter/lib/src/material/radio_list_tile.dart [hunk 7]
```diff
@@ -487,13 +487,13 @@ class RadioListTile<T> extends StatelessWidget {
           ),
         );
     }
-
+    final ThemeData theme = Theme.of(context);
+    final ListTileControlAffinity controlAffinity = this.controlAffinity ?? theme.listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/radio_list_tile.dart:491]:
```suggestion
    final ListTileThemeData listTileTheme = ListTileTheme.of(context);
    final ListTileControlAffinity effectiveControlAffinity = controlAffinity
      ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```

File: packages/flutter/lib/src/material/switch_list_tile.dart [hunk 8]
```diff
@@ -565,14 +565,14 @@ class SwitchListTile extends StatelessWidget {
           ),
         );
     }
-
+    final ThemeData theme = Theme.of(context);
+    final ListTileControlAffinity controlAffinity = this.controlAffinity ?? theme.listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/switch_list_tile.dart:569]:
```suggestion
    final ListTileThemeData listTileTheme = ListTileTheme.of(context);
    final ListTileControlAffinity effectiveControlAffinity = controlAffinity
      ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;     
```

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 9]
```diff
@@ -660,8 +660,9 @@ class _ExpansionTileState extends State<ExpansionTile> with SingleTickerProvider
   }
 
   // Platform or null affinity defaults to trailing.
-  ListTileControlAffinity _effectiveAffinity(ListTileControlAffinity? affinity) {
-    switch (affinity ?? ListTileControlAffinity.trailing) {
+  ListTileControlAffinity _effectiveAffinity() {
+    final ListTileControlAffinity affinity = widget.controlAffinity ?? Theme.of(context).listTileTheme.controlAffinity ?? ListTileControlAffinity.trailing;
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/expansion_tile.dart:664]:
```suggestion
    final ListTileThemeData listTileTheme = ListTileTheme.of(context);
    final ListTileControlAffinity affinity = widget.controlAffinity ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.trailing;
```

File: packages/flutter/test/material/checkbox_list_tile_test.dart [hunk 10]
```diff
@@ -1231,6 +1231,38 @@ void main() {
...
+            child: Material(
+              child: CheckboxListTile(
+                value: false,
+                onChanged: (bool? value) {},
+              ),
+            ),
+          );
+        }),
+      );
+    }
```
Comment (TahaTesser) [inline packages/flutter/test/material/checkbox_list_tile_test.dart:1254]:
```suggestion
    Widget buildListTile(ListTileControlAffinity controlAffinity) {
      return MaterialApp(
        theme: ThemeData(
          listTileTheme: ListTileThemeData(
            controlAffinity: controlAffinity,
          ),
        ),
        home: Material(
          child: CheckboxListTile(
            value: false,
            onChanged: (bool? value) {},
          ),
        ),
      );
    }
```

Please update this structure here and similar in other test classes

File: packages/flutter/test/material/checkbox_list_tile_test.dart [hunk 11]
```diff
@@ -1231,6 +1231,38 @@ void main() {
...
+
+    await tester.pumpWidget(buildView(ListTileControlAffinity.leading));
```
Comment (TahaTesser) [inline packages/flutter/test/material/checkbox_list_tile_test.dart:1256]:
```suggestion
    await tester.pumpWidget(buildListTile(ListTileControlAffinity.leading));
```

File: packages/flutter/test/material/checkbox_list_tile_test.dart [hunk 12]
```diff
@@ -1231,6 +1231,38 @@ void main() {
...
+    final Finder leading = find.byType(Checkbox);
+    final Offset offsetLeading = tester.getTopLeft(leading);
+    expect(offsetLeading, const Offset(16.0, 8.0));
+
+    await tester.pumpWidget(buildView(ListTileControlAffinity.trailing));
```
Comment (TahaTesser) [inline packages/flutter/test/material/checkbox_list_tile_test.dart:1261]:
```suggestion
    await tester.pumpWidget(buildListTile(ListTileControlAffinity.trailing));
    await tester.pumpAndSettle();

```

File: packages/flutter/test/material/checkbox_list_tile_test.dart (hunk 11 above)
Comment (TahaTesser) [reply packages/flutter/test/material/checkbox_list_tile_test.dart:1256]: Similar format in other test classes

File: packages/flutter/test/material/checkbox_list_tile_test.dart (hunk 12 above)
Comment (TahaTesser) [reply packages/flutter/test/material/checkbox_list_tile_test.dart:1261]: Similar format in other test classes. Add extra line like this one.

**Review by TahaTesser — CHANGES_REQUESTED** [review]
Thanks for the contribution!

**Comment by hello-coder-xu** [comment]
I updated it, please check again, thank you

File: packages/flutter/lib/src/material/checkbox_list_tile.dart [hunk 13]
```diff
@@ -518,10 +518,13 @@ class CheckboxListTile extends StatelessWidget {
         );
     }
 
-    final (Widget? leading, Widget? trailing) = switch (controlAffinity) {
+    final ListTileThemeData listTileTheme = ListTileTheme.of(context);
+    final ListTileControlAffinity effectiveControlAffinity = controlAffinity ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/checkbox_list_tile.dart:522]:
```suggestion
    final ListTileControlAffinity effectiveControlAffinity = controlAffinity ??
      listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```

`effectiveControlAffinity` are bit long for single. we can split them in all classes

**Review by TahaTesser — CHANGES_REQUESTED** [review]
This is almost good to go

File: packages/flutter/lib/src/material/checkbox_list_tile.dart (hunk 13 above)
Comment (hello-coder-xu) [reply packages/flutter/lib/src/material/checkbox_list_tile.dart]: I reformatted the code

File: packages/flutter/lib/src/material/radio_list_tile.dart [hunk 14]
```diff
@@ -488,11 +488,15 @@ class RadioListTile<T> extends StatelessWidget {
         );
     }
 
+    final ListTileThemeData listTileTheme = ListTileTheme.of(context);
+    final ListTileControlAffinity effectiveControlAffinity =
+        this.controlAffinity ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/radio_list_tile.dart:493]:
```suggestion
        controlAffinity ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```

We should avoid using `this`

File: packages/flutter/lib/src/material/switch_list_tile.dart [hunk 15]
```diff
@@ -566,6 +566,9 @@ class SwitchListTile extends StatelessWidget {
         );
     }
 
+    final ListTileThemeData listTileTheme = ListTileTheme.of(context);
+    final ListTileControlAffinity controlAffinity =
+        this.controlAffinity ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/switch_list_tile.dart:571]:
```suggestion
        controlAffinity ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```

File: packages/flutter/lib/src/material/checkbox_list_tile.dart [hunk 16]
```diff
@@ -518,10 +518,14 @@ class CheckboxListTile extends StatelessWidget {
         );
     }
 
-    final (Widget? leading, Widget? trailing) = switch (controlAffinity) {
+    final ListTileThemeData listTileTheme = ListTileTheme.of(context);
+    final ListTileControlAffinity effectiveControlAffinity =
+        controlAffinity ?? listTileTheme.controlAffinity ?? ListTileControlAffinity.platform;
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/checkbox_list_tile.dart:523]: This is good.

**Review by TahaTesser — CHANGES_REQUESTED** [review]
LGTM!

File: packages/flutter/lib/src/material/radio_list_tile.dart (hunk 14 above)
Comment (hello-coder-xu) [reply packages/flutter/lib/src/material/radio_list_tile.dart]: updated

**Review by TahaTesser — APPROVED** [review]
LGTM!

File: packages/flutter/test/material/switch_list_tile_test.dart [hunk 17]
```diff
@@ -1691,4 +1691,38 @@ void main() {
     expect(Focus.of(firstChildKey.currentContext!).hasPrimaryFocus, isFalse);
     expect(Focus.of(secondChildKey.currentContext!).hasPrimaryFocus, isTrue);
   });
+
+  testWidgets('SwitchListTile use ListTileTheme controlAffinity', (WidgetTester tester) async {
```
Comment (bleroux) [inline packages/flutter/test/material/switch_list_tile_test.dart:1695]: use -> uses

File: packages/flutter/test/material/radio_list_tile_test.dart [hunk 18]
```diff
@@ -1547,4 +1547,39 @@ void main() {
       );
     });
   });
+
+  testWidgets('RadioListTile use ListTileTheme controlAffinity', (WidgetTester tester) async {
```
Comment (bleroux) [inline packages/flutter/test/material/radio_list_tile_test.dart:1551]: use -> uses

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 19]
```diff
@@ -1585,4 +1585,36 @@ void main() {
 
     expect(titleSize.width, materialAppSize.width - 32.0);
   });
+
+  testWidgets('ExpansionTile use ListTileTheme controlAffinity', (WidgetTester tester) async {
```
Comment (bleroux) [inline packages/flutter/test/material/expansion_tile_test.dart:1589]: use -> uses

File: packages/flutter/test/material/checkbox_list_tile_test.dart [hunk 20]
```diff
@@ -1231,6 +1231,39 @@ void main() {
     expect(Focus.of(firstChildKey.currentContext!).hasPrimaryFocus, isFalse);
     expect(Focus.of(secondChildKey.currentContext!).hasPrimaryFocus, isTrue);
   });
+
+  testWidgets('CheckboxListTile use ListTileTheme controlAffinity', (WidgetTester tester) async {
```
Comment (bleroux) [inline packages/flutter/test/material/checkbox_list_tile_test.dart:1235]: use -> uses

File: packages/flutter/lib/src/material/list_tile_theme.dart [hunk 21]
```diff
@@ -462,6 +480,13 @@ class ListTileTheme extends InheritedTheme {
   /// [ListTileThemeData.enableFeedback] property instead.
   bool? get enableFeedback => _data != null ? _data.enableFeedback : _enableFeedback;
 
+  /// Overrides the default value of [CheckboxListTile.controlAffinity]
+  ///  or [ExpansionTile.controlAffinity] or [SwitchListTile.controlAffinity] or [RadioListTile.controlAffinity]
```
Comment (bleroux) [inline packages/flutter/lib/src/material/list_tile_theme.dart:484]: Remove extra space before 'or'

File: packages/flutter/lib/src/material/list_tile_theme.dart [hunk 22]
```diff
@@ -127,6 +128,10 @@ class ListTileThemeData with Diagnosticable {
   /// If specified, overrides the default value of [ListTile.titleAlignment].
   final ListTileTitleAlignment? titleAlignment;
 
+  /// If specified, overrides the default value of [CheckboxListTile.controlAffinity]
+  ///  or [ExpansionTile.controlAffinity] or [SwitchListTile.controlAffinity] or [RadioListTile.controlAffinity].
```
Comment (bleroux) [inline packages/flutter/lib/src/material/list_tile_theme.dart:132]: Remove extra space before 'or'

File: packages/flutter/lib/src/material/list_tile_theme.dart [hunk 23]
```diff
@@ -462,6 +480,13 @@ class ListTileTheme extends InheritedTheme {
   /// [ListTileThemeData.enableFeedback] property instead.
   bool? get enableFeedback => _data != null ? _data.enableFeedback : _enableFeedback;
 
+  /// Overrides the default value of [CheckboxListTile.controlAffinity]
+  ///  or [ExpansionTile.controlAffinity] or [SwitchListTile.controlAffinity] or [RadioListTile.controlAffinity]
+  ///
+  /// This property is obsolete: please use the [data]
```
Comment (bleroux) [inline packages/flutter/lib/src/material/list_tile_theme.dart:486]: Is the reference to 'data' valid here?

**Review by bleroux — COMMENTED** [review]
Thanks for the contribution, I spotted some typos

File: packages/flutter/lib/src/material/list_tile_theme.dart (hunk 23 above)
Comment (hello-coder-xu) [reply packages/flutter/lib/src/material/list_tile_theme.dart:486]: Yes. I tested it locally for [CheckboxListTile] [ExpansionTile] [SwitchListTile] [RadioListTile] and it works.

```
testWidgets('RadioListTile uses ListTileTheme controlAffinity', (WidgetTester tester) async {
    Widget buildListTile(ListTileControlAffinity controlAffinity) {
      return MaterialApp(
        home: Material(
          child: ListTileTheme(
            // data: ListTileThemeData(
            //   controlAffinity: controlAffinity,
            // ),
            controlAffinity: controlAffinity,
            child: RadioListTile<double>(
              value: 0.5,
              groupValue: 1.0,
              title: const Text('RadioListTile'),
              onChanged: (double? value) {},
            ),
          ),
        ),
      );
    }

    await tester.pumpWidget(buildListTile(ListTileControlAffinity.leading));
    final Finder leading = find.text('RadioListTile');
    final Offset offsetLeading = tester.getTopLeft(leading);
    expect(offsetLeading, const Offset(72.0, 16.0));

    await tester.pumpWidget(buildListTile(ListTileControlAffinity.trailing));
    final Finder trailing = find.text('RadioListTile');
    final Offset offsetTrailing = tester.getTopLeft(trailing);
    expect(offsetTrailing, const Offset(16.0, 16.0));

    await tester.pumpWidget(buildListTile(ListTileControlAffinity.platform));
    final Finder platform = find.text('RadioListTile');
    final Offset offsetPlatform = tester.getTopLeft(platform);
    expect(offsetPlatform, const Offset(72.0, 16.0));
  });
```

Do you think I should continue to add test cases for `data`?

Comment (bleroux) [reply packages/flutter/lib/src/material/list_tile_theme.dart:486]: Sorry for the confusion: my points was only on the formulation.

```dart

/// This property is obsolete: please use the [data]
/// [ListTileThemeData.controlAffinity] property instead.
```

Does it mean the following? :

```
/// This property is obsolete: please use the
/// [ListTileThemeData.controlAffinity] property instead.
```

**Review by bleroux — APPROVED** [review]
LGTM!

**Pull request merged**